gflownet/utils/batch.py

```diff
@@ -0,0 +1,179 @@
+from collections import defaultdict
+from copy import deepcopy
+
+import numpy as np
+import torch
+
+from gflownet.utils.common import tbool, tfloat, tint, tlong
+
+
+class Batch:
+    """
+    Important note: one env should correspond to only one trajectory, all env_id should be unique
+    """
+
+    def __init__(self, loss, device, float):
+        self.loss = loss
+        self.device = device
+        self.float = float
+        self.envs = dict()
+        self.state = []
+        self.action = []
+        self.done = []
+        self.env_id = []
+        self.mask_invalid_actions_forward = []
+        self.mask_invalid_actions_backward = []
+        self.parents = []
+        self.parents_actions = []
+        self.step = []
+
+    def __len__(self):
+        return len(self.state)
+
+    def add_to_batch(self, envs, actions, valids, train=True):
+        for env, action, valid in zip(envs, actions, valids):
+            self.envs.update({env.id: env})
+            if not valid:
+                continue
+            if train:
+                self.state.append(deepcopy(env.state))
+                self.action.append(action)
+                self.env_id.append(env.id)
+                self.done.append(env.done)
+                self.step.append(env.n_actions)
+                mask_f = env.get_mask_invalid_actions_forward()
+                self.mask_invalid_actions_forward.append(mask_f)
+                if self.loss == "flowmatch":
+                    parents, parents_a = env.get_parents(action=action)
+                    assert (
+                        action in parents_a
+                    ), f"""
+                    Sampled action is not in the list of valid actions from parents.
+                    \nState:\n{env.state}\nAction:\n{action}
+                    """
+                    self.parents.append(parents)
+                    self.parents_actions.append(parents_a)
+                if self.loss == "trajectorybalance":
+                    mask_b = env.get_mask_invalid_actions_backward(
+                        env.state, env.done, [action]
+                    )
+                    self.mask_invalid_actions_backward.append(mask_b)
+            else:
+                if env.done:
+                    self.state.append(env.state)
+                    self.env_id.append(env.id)
+                    self.step.append(env.n_actions)
+
+    def process_batch(self):
```

Comment:
This (and all of the others, to be honest) could use some docstrings with a short description of what's being done here.